src/ripple/app/hook/impl/applyHook.cpp

```diff
@@ -0,0 +1,4976 @@
+
+    // zero pad on the left
+    for (; i < pad; ++i)
+        key_buffer[i] = 0;
+
+    const char* data = source.data();
+
+    for (; i < 32; ++i)
+        key_buffer[i] = data[i - pad];
+
+    return ripple::uint256::fromVoid(key_buffer);
+}
+
+// check the state cache
+inline
+std::optional<std::reference_wrapper<std::pair<bool, ripple::Blob> const>>
+lookup_state_cache(
+        hook::HookContext& hookCtx,
+        ripple::AccountID const& acc,
+        ripple::uint256 const& ns,
+        ripple::uint256 const& key)
+{
+    std::cout << "Lookup_state_cache: acc: " << acc << " ns: " << ns << " key: " << key << "\n";
+    auto& stateMap = hookCtx.result.stateMap;
+    if (stateMap.find(acc) == stateMap.end())
+        return std::nullopt;
+
+    auto& stateMapAcc = stateMap[acc].second;
+    if (stateMapAcc.find(ns) == stateMapAcc.end())
+        return std::nullopt;
+
+    auto& stateMapNs = stateMapAcc[ns];
+
+    auto const& ret = stateMapNs.find(key);
+
+    if (ret == stateMapNs.end())
+        return std::nullopt;
+    
+    return std::cref(ret->second);
+}
+
+
+// update the state cache
+// RH TODO: add an accumulator for newly reserved state to the state map so canReserveNew can be computed
+// correctly when more than one new state is reserved.
+inline
+bool                                    // true unless a new hook state was required and the acc has insufficent reserve
+set_state_cache(
+        hook::HookContext& hookCtx,
+        ripple::AccountID const& acc,
+        ripple::uint256 const& ns,
+        ripple::uint256 const& key,
+        ripple::Blob& data,
+        bool modified)
+{
+
+    auto& stateMap = hookCtx.result.stateMap;
+    if (stateMap.find(acc) == stateMap.end())
+    {
+
+        // if this is the first time this account has been interacted with
+        // we will compute how many available reserve positions there are
+        auto const& fees = hookCtx.applyCtx.view().fees();
+
+        auto const accSLE = hookCtx.applyCtx.view().read(ripple::keylet::account(acc));
+        if (!accSLE)
+            return false;
+
+        STAmount bal = accSLE->getFieldAmount(sfBalance);
+
+        int64_t availableForReserves =
+            bal.xrp().drops() - fees.accountReserve(accSLE->getFieldU32(sfOwnerCount)).drops();
+
+        int64_t increment = fees.increment.drops();
+
+        if (increment <= 0)
+            increment = 1;
+
+        availableForReserves /= increment;
+
+        if (availableForReserves < 1 && modified)
+            return false;
+        
+        stateMap[acc] =
+        {
+            availableForReserves - 1,
+            {
+                {
+                    ns,
+                    {
+                        {
+                            key,
+                            {
+                               modified,
+                               data
+                            }
+                        }
+                    }
+                }
+            }
+        };
```

Comment:
For reference the type is below. We could hide this away in a wrapper class instead of exposing it directly, but as far as maps of maps go... I think you always end up with something that looks like this. Open to suggestions/ideas. Maybe it can be flattened somehow?
```c++
    // This map type acts as both a read and write cache for hook execution
    // and is preserved across the execution of the set of hook chains
    // being executed in the current transaction. It is committed to lgr
    // only upon tesSuccess for the otxn.
    using HookStateMap =
        std::map<
            ripple::AccountID,                  // account that owns the state
            std::pair<
                int64_t,                        // remaining available ownercount
                std::map<ripple::uint256,       // namespace
                std::map<ripple::uint256,       // key
                std::pair<
                    bool,                       // is modified from ledger value
                    ripple::Blob>>>>>;          // the value
```